### PR TITLE
Add test for async client certificate validation.

### PR DESCRIPTION
**Pull request by anrossi**
## Description

Add test to repro scenario where server asynchronously validates a client certificate.
Update `ConnectionCertificateValidationComplete` to allow calling on server connections as well.

## Testing

Existing and new tests in CI.

## Documentation

Update the documentation for `ConnectionCertificateValidationComplete` to allow use for server scenarios.

### CODE REVIEW

**Comment by nibanks** [comment]
```
/home/vsts/work/1/s/src/test/lib/HandshakeTest.cpp: In function ‘void QuicTestCustomClientCertificateValidation(bool, bool)’:
/home/vsts/work/1/s/src/test/lib/HandshakeTest.cpp:879:52: error: invalid conversion from ‘void (*)(TestConnection*, HQUIC, QUIC_STREAM_OPEN_FLAGS)’ {aka ‘void (*)(TestConnection*, QUIC_HANDLE*, QUIC_STREAM_OPEN_FLAGS)’} to ‘void*’ [-fpermissive]
  879 |                 ServerAcceptCtx.NewStreamHandler = NewStreamCallbackTestFail;
      |                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                    |
      |                                                    void (*)(TestConnection*, HQUIC, QUIC_STREAM_OPEN_FLAGS) {aka void (*)(TestConnection*, QUIC_HANDLE*, QUIC_STREAM_OPEN_FLAGS)}
```

**Review by nibanks — DISMISSED** [review]

**Comment by nibanks** [comment]
Looks like the tests are failing:
```
[ RUN      ] Handshake/WithHandshakeArgs6.ConnectClientCertificate/1
D:\a\1\msquic\src\test\lib\TestConnection.cpp(810): error: Unexpected transport Close Error, expected=0x0, actual=0x8041012a
D:\a\1\msquic\src\test\lib\HandshakeTest.cpp(2459): error: Server->GetIsConnected() not equal to UseClientCertificate
D:\a\1\msquic\src\test\lib\TestConnection.cpp(810): error: Unexpected transport Close Error, expected=0x0, actual=0x8041012a
[  FAILED  ] Handshake/WithHandshakeArgs6.ConnectClientCertificate/1, where GetParam() = v4/Cert (57 ms)
[----------] 1 test from Handshake/WithHandshakeArgs6 (57 ms total)
```

File: src/test/lib/HandshakeTest.cpp [hunk 1]
```diff
@@ -820,6 +821,125 @@ QuicTestCustomCertificateValidation(
...
+
+                if (!CxPlatEventWaitWithTimeout(ServerAcceptCtx.NewConnectionReady, TestWaitTimeout)) {
+                    TEST_FAILURE("Timed out waiting for server accept.");
+                }
+
+                if (AsyncValidation) {
+                    CxPlatSleep(2000);
+                    BOOLEAN Value = !!AcceptCert;
+                    TEST_QUIC_SUCCEEDED(
+                        MsQuic->SetParam(
```
[the remark this answers is not in the document]
Comment (anrossi) [reply src/test/lib/HandshakeTest.cpp:919]: The `SetParam` call still works for server side.  Is there a reason that `MsQuicConnectionCertificateValidationComplete` is client-only?

[the remark this answers is not in the document]
Comment (ami-GS) [reply src/test/lib/HandshakeTest.cpp:924]: I simply thought the validation is for client only feature. I believe I saw such code path. maybe?
It is ok to remove the assertion from api.c

[the remark this answers is not in the document]
Comment (ami-GS) [reply src/test/lib/HandshakeTest.cpp:924]: If you use, could you please fix document says client only
https://github.com/microsoft/msquic/blob/main/docs/api/ConnectionCertificateValidationComplete.md

**Review by nibanks — APPROVED** [review]

**Pull request merged**